Feature/forhighres

JinsuRhee opened the pull request:
* Node skip & closing algorithms are added (Currently, just for FOFSearchBall & FOFSearchCriterion routines but it's also possible to apply the new logics to other FOF routines).

* KDTree building routines are added (for OMP decomposition, for adaptive trees, and for FOFSearchCriterion).

* Some bug(?) fixed. Please check whether it is correctly revised.

pelahi committed:
Update to split nodes, allow particle tagging at split node level

JinsuRhee committed:
Add Node Closing Algorithm ## If its all son nodes are closed, the split node get closed

JinsuRhee committed:
Leaf Node Tagging added

JinsuRhee committed:
Find Leaf node by its tag value

JinsuRhee committed:
(Bugs?) Wrong calculation for rd ## In 6D case, invscaling value can be different for old_off

JinsuRhee committed:
(Bugs?) Is it truely needed to be Group[id]==iGroup rather than Group[id] ?

JinsuRhee committed:
Build Nodes for the OMP decomposition added by using adaptive KDTree && related Qsort routine added

JinsuRhee committed:
KDTree for building OMP domains

JinsuRhee committed:
Build Nodes for a normal adaptive KDTree && Node Skip Parameter routine added

JinsuRhee committed:
KDTree for building normal adaptive KDTree

JinsuRhee committed:
Node Closing & Skipping algorithm added in the FOFSearchBall routine

JinsuRhee committed:
Build Nodes for the FOFSearchCriterion routine

JinsuRhee committed:
KDTree for the FOFSearchCriterion routine

JinsuRhee committed:
Node Closing & Skipping algorithm added in the FOFSearchCriterion routine

JinsuRhee committed:
Node Closing & Skipping algorithm added in the FOFSearchCriterion routine

JinsuRhee committed:
Merge branch 'feature/forhighres' of github.com:JinsuRhee/NBodylib into HEAD

pelahi commented:
Hi @JinsuRhee, I have some questions regarding your updates. 

- I am trying to understand some of the extra KDTree build routines you have implemented. First, you have added BuildNodes_ADT, which I understand is the adaptive tree with the minimum sphere calculation designed to end nodes at the leaf node that has a smaller sphere that the search radius. That is a tree construction optimised for nD FOF searches (here so long as dimensions are appropriately scaled to calculate |x| simply. That seems fine, though given the extra compute I wonder if it is really any better than a simple KDTree with a split node closing criterion added as I have done. **Thus, the first question is does the extra computation really speed up the FOF algorithm if split nodes can close themselves? Have you tried to see how long the new feature branch I created is compared to your branch?** 
- The next addition I think is formally incorrect. There is no way to generalise the FOFCriterion in the fashion you have done simply because you have no idea what criterion will be imposed. Can you explain your motivation for adding this function?

JinsuRhee commented:
Hi @pelahi 

For the first question, actually, that is for the calculation of the farthest particle in each node to determine whether that node is entirely enclosed within the search radius. I know the additional computations make building trees get slower. Nevertheless, the advantage of using additional calculations are, I think, 

1) Do not depend on the physical shape of a node. The previous logic is to use vertex points of a node for the above purpose, which is sometimes not effective if the node shape is very irregular.
2) Can skip that node if the node is entirely outside the search radius.

I haven't tested, however, how fast the new node skip scheme is compared to the original version. I'll test it soon.

For the second question, I don't know which addition you mean. Could you specify it?

pelahi committed:
Update to API and allow for more intuitive adaptive tree

This is a proposed update to change the API so that a single KDTree constructor can be used to build an adaptive tree. There are also some name changes to internal variables and a proposed clean-up to some code. I will make further changes in later commits

pelahi committed:
Fix Find leaf functions to handle bool

pelahi committed:
Bug fixes for particle comparison functions 

Fix is to correctly use std::sort

pelahi committed:
Update to KDNodes, how particles searched for FOF and searchball

A revision to FOF search to quickly tag all particles if node is inside search radius. Also update to return FOFCriterion searches to non-optimised ones. Suggestion is to implement new FOF search with clearly defined API indicating simple nD FOF search but where perhaps there is criterion for type of particle that can form links. Generalised FOF searches do not have trivial closing criteria.

pelahi committed:
Here removed extra FOF constructors

Now default constructor has additional argument that can be passed to enable adaptive tree construction.

pelahi committed:
Added OpenMP parallelism to center and farthest calculation

pelahi committed:
Remove unnecessary BuildNodes

pelahi committed:
Bug fix to handle phase-space trees

Fixed DetermineCentreAndSmallestSphere to handle phase-space distances

pelahi committed:
Update to leaf and split nodes for full node tagging

This update adds node closing criterion to SearchBallPos functions and FOF searches. The FOF search check using the node center and distance to furtherst particle needs some more thought to see if it works.

pelahi committed:
Added function for approximative median tree

pelahi committed:
Added function for approximative median tree

pelahi committed:
Added function for approximative median tree

Added function for approximative median tree

pelahi committed:
Merge remote-tracking branch 'origin/feature/pelahi-forhighrest' into feature/pelahi-forhighrest

pelahi committed:
Adaptive tree update

Tree will now build till either leaf node is smaller than some size or leaf node contains fewer than some number of particles

pelahi committed:
Update to adaptive tree construction

Now allow for non median split if the AdaptiveMedianFac argument in tree construction is nonzero.

JinsuRhee committed:
Merge pull request #1 from JinsuRhee/feature/pelahi-forhighrest

Feature/pelahi forhighrest

JinsuRhee committed:
Incorrect tagging fixed and node skip scheme added

JinsuRhee committed:
Incorrect tagging fixed and node skip scheme added

JinsuRhee committed:
Run error in DetermineCentreandSmallestSphere when OMP turn on

JinsuRhee committed:
Buffer size is added to adaptive tree build & bugs in it are fixed

pelahi committed:
Updates for adaptive trees

Have updated to have a opening a node criterion check, fixed some bugs in the adaptive tree construction and entropy based trees.

pelahi committed:
Merge remote-tracking branch 'origin/feature/pelahi-forhighrest' into feature/pelahi-forhighrest

JinsuRhee committed:
Merge pull request #3 from JinsuRhee/feature/pelahi-forhighrest

Feature/pelahi forhighrest

JinsuRhee committed:
Revert "Feature/pelahi forhighrest"

JinsuRhee committed:
Merge pull request #4 from JinsuRhee/revert-3-feature/pelahi-forhighrest

Revert 3 feature/pelahi forhighrest

pelahi committed:
Revert unintended changes

Undo incorrectly pushed changes merged from another branch

pelahi committed:
Updates to adaptive tree building

Added the option to continue splitting if interparticle spacing greater than desired maximal leaf node size. Also fixed an OpenMP bug when determining the centre of mass and the distance to the farthest particle of nodes.

pelahi committed:
OpenMP bug fix

pelahi committed:
Bug fix for running adaptive tree with interparticle spacing splitting criterion

pelahi committed:
Merge branch 'feature/splitnodeoptmisation' into feature/pelahi-forhighrest

pelahi committed:
Merge pull request #6 from JinsuRhee/feature/pelahi-forhighrest

Feature/pelahi forhighrest

pelahi committed:
Bug fix

pelahi committed:
Reinstated incorrectly commented line

JinsuRhee committed:
(Bug fixed) off in FOF should be initalized to 6dim because the routine is also used in 6D search

pelahi committed:
Merge pull request #7 from JinsuRhee/feature/pelahi-forhighrest

(Bug fixed) off in FOF should be initialized to 6dim

JinsuRhee committed:
Incorrect range of the loop index
